gadgets/trace_ssl/program.bpf.c

```diff
@@ -0,0 +1,432 @@
+
+	mntns_id = gadget_get_mntns_id();
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return 0;
+
+	tid = (u32)bpf_get_current_pid_tgid();
+	ts = bpf_ktime_get_boot_ns();
+
+	bpf_map_update_elem(&mntns_context, &tid, &mntns_id, BPF_ANY);
+	bpf_map_update_elem(&ssl_starttime, &tid, &ts, BPF_ANY);
+	return 0;
+}
+
+SEC("uretprobe/libssl:SSL_do_handshake")
+int trace_uretprobe_libssl_SSL_do_handshake(struct pt_regs *ctx)
+{
+	struct event *event;
+	u64 *mntns_ptr;
+	u64 pid_tgid;
+	u64 uid_gid;
+	u32 tid;
+	u64 ts;
+	u64 *tsp;
+	int ret;
+
+	ts = bpf_ktime_get_boot_ns();
+	pid_tgid = bpf_get_current_pid_tgid();
+	uid_gid = bpf_get_current_uid_gid();
+	tid = (u32)pid_tgid;
+
+	tsp = bpf_map_lookup_elem(&ssl_starttime, &tid);
+	if (tsp == 0)
+		goto clean;
+
+	mntns_ptr = bpf_map_lookup_elem(&mntns_context, &tid);
+	if (mntns_ptr == 0)
+		goto clean;
+
+	ret = PT_REGS_RC(ctx);
+	if (ret <= 0) // handshake failed
+		goto clean;
+
+	event = gadget_reserve_buf(&events, BASE_EVENT_SIZE);
+	if (!event)
+		goto clean;
+
+	event->mntns_id = *mntns_ptr;
+	event->operation = libssl_SSL_do_handshake;
+	event->timestamp = ts;
+	event->latency_ns = ts - *tsp;
+	event->pid = pid_tgid >> 32;
+	event->tid = tid;
+	event->uid = uid_gid;
+	event->gid = uid_gid >> 32;
+	event->len = 0;
+	event->retval = PT_REGS_RC(ctx);
+	bpf_get_current_comm(&event->comm, sizeof(event->comm));
+
+	gadget_submit_buf(ctx, &events, event, BASE_EVENT_SIZE);
+
+clean:
+	bpf_map_delete_elem(&mntns_context, &tid);
+	bpf_map_delete_elem(&ssl_starttime, &tid);
+	return 0;
+}
+
+/* uprobes for libgnutls */
+PROBE_SSL_RW_ENTER(libgnutls, gnutls_record_send)
+PROBE_SSL_RW_EXIT(libgnutls, gnutls_record_send)
+
+PROBE_SSL_RW_ENTER(libgnutls, gnutls_record_recv)
+PROBE_SSL_RW_EXIT(libgnutls, gnutls_record_recv)
+
+/* uprobes for libnss */
+PROBE_SSL_RW_ENTER(libnss, PR_Write)
+PROBE_SSL_RW_EXIT(libnss, PR_Write)
+
+PROBE_SSL_RW_ENTER(libnss, PR_Send)
+PROBE_SSL_RW_EXIT(libnss, PR_Send)
+
+PROBE_SSL_RW_ENTER(libnss, PR_Read)
+PROBE_SSL_RW_EXIT(libnss, PR_Read)
+
+PROBE_SSL_RW_ENTER(libnss, PR_Recv)
+PROBE_SSL_RW_EXIT(libnss, PR_Recv)
+
+/* uprobes for libcrypto */
+static __always_inline int probe_crypto_enter(struct pt_regs *ctx)
+{
+	u64 mntns_id;
+	u32 tid;
+	u64 ts;
+
+	mntns_id = gadget_get_mntns_id();
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return 0;
+
+	tid = (u32)bpf_get_current_pid_tgid();
+	ts = bpf_ktime_get_boot_ns();
+
+	bpf_map_update_elem(&mntns_context, &tid, &mntns_id, BPF_ANY);
+	bpf_map_update_elem(&crypto_starttime, &tid, &ts, BPF_ANY);
+	return 0;
+}
+
+static __always_inline int probe_crypto_exit(struct pt_regs *ctx,
+					     enum operation op)
+{
+	struct event *event;
+	u64 *mntns_ptr;
+	u64 pid_tgid;
+	u64 uid_gid;
+	u32 tid;
+	u64 ts;
+	u64 *tsp;
+
+	ts = bpf_ktime_get_boot_ns();
+	pid_tgid = bpf_get_current_pid_tgid();
+	uid_gid = bpf_get_current_uid_gid();
+	tid = (u32)pid_tgid;
+
+	tsp = bpf_map_lookup_elem(&crypto_starttime, &tid);
+	if (tsp == 0)
```

Comment:
Since it is a pointer, we would usually write `NULL` instead of `0`, or just use:
```suggestion
	if (!tsp)
```